storage/transfermanager/downloader.go

```diff
@@ -17,61 +17,131 @@ package transfermanager
+// Downloader manages a set of parallelized downloads.
 type Downloader struct {
-	client  *storage.Client
-	config  *transferManagerConfig
-	work    chan *DownloadObjectInput // Piece of work to be executed.
-	output  chan *DownloadOutput      // Channel for completed downloads; used to feed results iterator.
-	workers *sync.WaitGroup           // Keeps track of the workers that are currently running.
+	client    *storage.Client
+	config    *transferManagerConfig
+	inputs    []DownloadObjectInput
+	results   []DownloadOutput
+	errors    []error
+	inputsMu  *sync.Mutex
+	resultsMu *sync.Mutex
+	errorsMu  *sync.Mutex
+	work      chan *DownloadObjectInput // Piece of work to be executed.
+	done      chan bool                 // Indicates to finish up work; expecting no more inputs.
+	workers   *sync.WaitGroup           // Keeps track of the workers that are currently running.
 }
 
-// DownloadObject queues the download of a single object. If it's larger than
-// the specified part size, the download will automatically be broken up into
-// multiple range reads. This will initiate the download but is non-blocking;
-// call Downloader.Results to process the result.
+// DownloadObject queues the download of a single object. This will initiate the
+// download but is non-blocking; call Downloader.Results to process the result.
 func (d *Downloader) DownloadObject(ctx context.Context, input *DownloadObjectInput) {
 	input.ctx = ctx
-	d.work <- input
+	d.addInput(input)
 }
 
-// Download a set of objects to a local path. Downloader will create the needed
-// directory structure locally as the operations progress.
-// This will initiate the download but is non-blocking; wait on Downloader.Results
-// to process the result. Results will be split into individual objects.
-// NOTE: Do not use, DownloadDirectory is not implemented.
-func (d *Downloader) DownloadDirectory(ctx context.Context, input *DownloadDirectoryInput) {
-	d.output <- &DownloadOutput{Bucket: input.Bucket, Err: errors.New("DownloadDirectory is not implemented")}
-	// This does obj listing/os.File calls and then converts to object inputs.
+// DownloadObjectWithCallback queues the download of a single object. This will
+// initiate the download but is non-blocking; use the callback to process the
+// result.
+// The results of downloads initiated with this method will not be provided in
+// the results given in Downloader.Results.
+func (d *Downloader) DownloadObjectWithCallback(ctx context.Context, input *DownloadObjectInput, callback func(*DownloadOutput)) {
+	input.ctx = ctx
+	input.callback = &callback
+	d.addInput(input)
 }
 
-// Waits for all outstanding downloads to complete. The Downloader must not be
-// used to download more objects or directories after this has been called.
+// WaitAndClose waits for all outstanding downloads to complete. The Downloader
+// must not be used for any more downloads after this has been called.
```

Comment:
We aren't enforcing it, but it should panic when trying to send on the closed channels. I can change the wording (and maybe also mention that it will panic) and add a test for it.

Reply:
I changed this to error instead of panicking.